mayastor/src/bdev/nexus/nexus_child_error_store.rs

```diff
@@ -0,0 +1,302 @@
+                self.records[idx].io_op,
+                self.records[idx].io_error,
+                self.records[idx].io_offset,
+                self.records[idx].io_num_blocks,
+            )
+            .expect("invalid format");
+        }
+        Ok(())
+    }
+}
+
+impl Debug for NexusErrStore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        self.error_fmt(f)
+    }
+}
+
+impl Display for NexusErrStore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        self.error_fmt(f)
+    }
+}
+
+impl Nexus {
+    pub fn error_record_add(
+        &self,
+        bdev: *const spdk_bdev,
+        io_op_type: spdk_bdev_io_type,
+        io_error_type: i32,
+        io_offset: u64,
+        io_num_blocks: u64,
+    ) {
+        let now = Instant::now();
+        let cfg = Config::by_ref();
+        if cfg.err_store_opts.enable_err_store {
+            let nexus_name = self.name.clone();
+            // dispatch message to management core to do this
+            let mgmt_reactor = Reactors::get_by_core(Cores::first()).unwrap();
+            mgmt_reactor.send_future(async move {
+                Nexus::future_error_record_add(
+                    nexus_name,
+                    bdev,
+                    io_op_type,
+                    io_error_type,
+                    io_offset,
+                    io_num_blocks,
+                    now,
+                );
+            });
+        }
+    }
+
+    fn future_error_record_add(
+        name: String,
+        bdev: *const spdk_bdev,
+        io_op_type: spdk_bdev_io_type,
+        io_error_type: i32,
+        io_offset: u64,
+        io_num_blocks: u64,
+        now: Instant,
+    ) {
+        let nexus = match nexus_lookup(&name) {
+            Some(nexus) => nexus,
+            None => {
+                error!("Failed to find the nexus {}", name);
+                return;
+            }
+        };
+        for child in nexus.children.iter_mut() {
+            if child.bdev.as_ref().unwrap().as_ptr() as *const _ == bdev {
+                if child.err_store.is_some() {
+                    child.err_store.as_mut().unwrap().add_record(
+                        io_op_type,
+                        io_error_type,
+                        io_offset,
+                        io_num_blocks,
+                        now,
+                    );
+                } else {
+                    error!("Failed to record error - child has no error store");
+                }
+                return;
+            }
+        }
+        error!("Failed to record error - could not find child");
+    }
+
+    pub fn error_record_query(
+        &self,
+        child_name: &str,
+        io_op_flags: u32,
+        io_error_flags: u32,
+        age_nano: u64,
+    ) -> Result<Option<u32>, nexus_bdev::Error> {
+        let earliest_time = Instant::now()
+            .checked_sub(Duration::from_nanos(age_nano))
+            .unwrap();
```

Comment:
should this unwrap be handled? eg if the time drifts?

Reply:
Hmm... the unwrap should be handled but not because of time drifting. The monotonic clock has an unspecified start time so it's possible that a large enough age_nano could "underflow" it.

Reply:
Ok, will do.